core/src/main/kotlin/io/specmatic/core/pattern/Discriminator.kt

```diff
@@ -164,32 +170,32 @@ class Discriminator(
 
         val deepMatchResults = failures.filter { it.hasReason(FailureReason.FailedButDiscriminatorMatched) }
 
-        return if(deepMatchResults.isNotEmpty())
-            Result.Failure.fromFailures(deepMatchResults)
-                .removeReasonsFromCauses().copy(failureReason = FailureReason.FailedButDiscriminatorMatched)
+        return if (deepMatchResults.isNotEmpty())
+            Result.Failure.fromFailures(deepMatchResults).removeReasonsFromCauses().copy(failureReason = FailureReason.FailedButDiscriminatorMatched)
         else
             Result.Failure.fromFailures(failures).removeReasonsFromCauses()
-
     }
 
     private fun discriminatorMatchFailure(pattern: Pattern) = AnyPattern.AnyPatternMatch(
         pattern,
         Result.Failure(
             "Discriminator match failure",
-            failureReason = FailureReason.DiscriminatorMismatch
+            failureReason = FailureReason.DiscriminatorMismatch,
+            ruleViolation = StandardRuleViolation.DISCRIMINATOR_MISMATCH
         )
     )
 
     private fun jsonObjectMismatchError(
         resolver: Resolver,
         sampleData: Value?
-    ) = resolver.mismatchMessages.valueMismatchFailure("json object", sampleData)
+    ) = valueMismatchResult("json object", sampleData, resolver.mismatchMessages)
 
     private fun discriminatorKeyMissingFailure(discriminatorProperty: String, discriminatorCsv: String) =
         Result.Failure(
             "Discriminator property $discriminatorProperty is missing from the object (it's value should be $discriminatorCsv)",
             breadCrumb = discriminatorProperty,
-            failureReason = FailureReason.DiscriminatorMismatch
+            failureReason = FailureReason.DiscriminatorMismatch,
```

Comment:
`failureReason` and `ruleViolation` appear very similar if you just look at the names. However, they serve different purposes, and maybe the names could make this clearer.